### PR TITLE
[fetch] Introduce test generation tool & add more tests for metadata request headers

### PR DESCRIPTION
**Pull request by jugglinmike**
This patch is the latest in a research project we've been running since the beginning of the year. The `README.md` file included in this patch summarizes the practical usage aspects and the high-level design goals.

If you're interested in learning more about the design process, you can check out:

- [Fetch Metadata research report](https://docs.google.com/document/d/1zyV4RzHnP8xqyxH4yHiNanrIoj7tjNYqQtligAgoPN4/edit#) - an initial survey of [the "Fetch Metadata request headers" proposal](https://w3c.github.io/webappsec-fetch-metadata/), including a review of interesting coverage areas
- [Fetch Metadata Test Generation Report](https://docs.google.com/document/d/1jOcNgUIF3OhCKptt1GbJLd--4CfTMc5cVrWEz6aso9s/edit) - a summary of some of the tool's evolution
- [Prototype for fetch metadata test generation](https://github.com/bocoup/wpt/pull/13) - the complete development and peer review process of the prototype

This patch replaces tests that were manually written with tests that have been procedurally generated by the new tool. I've verified that there are no regressions in coverage. To promote transparency on that detail, I've organized each replacement as a distinct commit in this pull request. If/when this is accepted for WPT, I recommend using a merge commit in order to preserve this information.

There are still more tests to be written (both to replace existing tests and to further improve coverage). I will be pushing additional commits to this feature branch over the next few days, though I do not expect to reach completeness. I'll be careful not to rewrite history to avoid disrupting the review process.

### CODE REVIEW

**Comment by jugglinmike** [comment]
Also CC @lweichselbaum (sorry Lukas: GitHub didn't allow me to designate you as a reviewer)

**Review by mikewest — APPROVED** [review]
I reviewed earlier iterations on this PR in detail at https://github.com/bocoup/wpt/pull/13, and I'm quite supportive of landing it. Thank you for the hard work you've done to get this into reasonable shape, @jugglinmike! LGTM.

**Comment by jugglinmike** [comment]
The check named, "wpt-decision-task" failed on the latest commit, reporting:

    subprocess.CalledProcessError: Command '[u'git', 'diff', '--name-only', '-z', u'0d794eb31a9920dd63065a1c4de2d14132b986f5..5e4ad738d4758f62d16410afae6485f839cf8f44']' returned non-zero exit status 128
    [taskcluster 2020-08-28 18:39:39.241Z] === Task Finished ===
    [taskcluster 2020-08-28 18:39:39.336Z] Unsuccessful task run with exit code: 1 completed in 45.22 seconds

It had a similar problem with the prior head of this branch, though it was successful when I initially opened this pull request. Both of the failed commands work locally, so this may be related to branch availability. This branch is located on a fork of WPT.

@Hexcles (hello!) are you familiar with the failing task?

**Comment by Hexcles** [comment]
I saw a similar failure on my PR just now but forgot to keep the link. Closing and reopening the PR seems to have fixed it.

**Comment by jugglinmike** [comment]
Thanks; I'll give that a try

**Review by arturjanc — APPROVED** [review]
Thanks for doing this, Mike! 

I really like the readability of the config format in fetch/metadata/tools/fetch-metadata.conf.yml.

**Comment by jugglinmike** [comment]
I'm glad to hear that, @arturjanc. The configuration file is where the complexity of the generation logic is most apparent, so I really tried to make it as clear as possible. There's still a lot to take in, but it's encouraging to hear that someone other than me can make some sense of it.

**Comment by zcorpan** [comment]
From what I understand, this is reviewed and ready to land. The stability checks failed I believe because this introduces too many tests and so the stability checks time out.

@jugglinmike is currently working on a different project. Worst case if these tests somehow catch fire, we can revert.

Any objection to merging?

**Comment by mikewest** [comment]
Ha! I thought this landed months ago. Yes. Please merge it. :)

**Comment by zcorpan** [comment]
> This patch replaces tests that were manually written with tests that have been procedurally generated by the new tool. I've verified that there are no regressions in coverage. To promote transparency on that detail, I've organized each replacement as a distinct commit in this pull request. If/when this is accepted for WPT, I recommend using a merge commit in order to preserve this information.

Merge commits are disabled for this repo. But we can "rebase and merge" without squashing into a single commit.

**Comment by zcorpan** [comment]
Hmmmmmm, this looks problematic:

```
+ python fetch/metadata/tools/generate.py
Traceback (most recent call last):
  File "fetch/metadata/tools/generate.py", line 7, in <module>
    import jinja2
ImportError: No module named jinja2
```

https://community-tc.services.mozilla.com/tasks/EMh-wimiQ7qr608Mw9cITw/runs/0/logs/https%3A%2F%2Fcommunity-tc.services.mozilla.com%2Fapi%2Fqueue%2Fv1%2Ftask%2FEMh-wimiQ7qr608Mw9cITw%2Fruns%2F0%2Fartifacts%2Fpublic%252Flogs%252Flive.log#L4047

**Comment by zcorpan** [comment]
I tried to fix the error by adding a `build.sh` script that installs dependencies and runs `generate.py` correctly within a virtualenv, which works locally but still fails in CI, not sure why: https://community-tc.services.mozilla.com/tasks/AnXTHQv7RDmXRIOPHZyRdQ/runs/0/logs/https%3A%2F%2Fcommunity-tc.services.mozilla.com%2Fapi%2Fqueue%2Fv1%2Ftask%2FAnXTHQv7RDmXRIOPHZyRdQ%2Fruns%2F0%2Fartifacts%2Fpublic%252Flogs%252Flive.log

I thought a rebase again might help, but https://github.com/web-platform-tests/wpt/pull/25247/commits/1099de7d07e403c83a3b7f78fa754d895873ed26 now conflicts with this change: https://github.com/web-platform-tests/wpt/commit/bcb6b4206fa9bbeedd20cae1f17593a0915f9ddd (fyi @smaug---- ), which looks like it might need some adjustment to the history test to not lose that change.

**Comment by zcorpan** [comment]
Per @jgraham , the previous error wasn't my fault, and now it's failing because there is a merge conflict. So next step is to fix the merge conflict and then try again.

**Comment by zcorpan** [comment]
> I thought a rebase again might help, but [1099de7](https://github.com/web-platform-tests/wpt/commit/1099de7d07e403c83a3b7f78fa754d895873ed26) now conflicts with this change: [bcb6b42](https://github.com/web-platform-tests/wpt/commit/bcb6b4206fa9bbeedd20cae1f17593a0915f9ddd) (fyi @smaug---- ), which looks like it might need some adjustment to the history test to not lose that change.

I looked at these commits now and discussed with @jugglinmike. @smaug---- 's change is a bug fix for test stability, which @jugglinmike had also fixed in essentially the same way in his rewrite. Namely, to use the `load` event instead of a timeout to make sure the navigation results in a new history entry being added. No new assertions or anything else was added in [bcb6b42](https://github.com/web-platform-tests/wpt/commit/bcb6b4206fa9bbeedd20cae1f17593a0915f9ddd)

Therefore, we can rebase and throw away that change without losing anything.

**Comment by zcorpan** [comment]
I've now rebased this on current master and resolved the conflict by removing [bcb6b42](https://github.com/web-platform-tests/wpt/commit/bcb6b4206fa9bbeedd20cae1f17593a0915f9ddd).

```
	new file:   fetch/metadata/generated/window-history.https.sub.html
	new file:   fetch/metadata/generated/window-history.sub.html
	deleted:    fetch/metadata/history.https.sub.html
	deleted:    fetch/metadata/resources/go-back.html
	deleted:    fetch/metadata/resources/post-to-owner-onload.py
	new file:   fetch/metadata/tools/templates/window-history.sub.html
```

The previous state of this branch is still available here: https://github.com/bocoup/wpt/tree/fetch-metadata-generation-2021-01-11

**Comment by zcorpan** [comment]
`wpt-firefox-nightly-stability` and `wpt-chrome-dev-stability` failed because they timed out, I believe because this PR adds so many tests.

The other issues are resolved, so this can be admin-merged. I could merge myself technically, but would like to ask @web-platform-tests/admins for a sign-off.

**Comment by stephenmcgruer** [comment]
> The other issues are resolved, so this can be admin-merged. I could merge myself technically, but would like to ask @web-platform-tests/admins for a sign-off.

Happy to do so. Should this be rebase-merged given the long commit history?

I have one other question: are the affected-tests 'regressions' from Chrome + Firefox expected? (Both see a few tests go from `X/X` to `0/1`, which in my experience usually means a harness error!)

**Comment by mikewest** [comment]
Perhaps we could run this through Chromium's commit queue to hammer out those issues?

**Comment by zcorpan** [comment]
@stephenmcgruer 

> Happy to do so. Should this be rebase-merged given the long commit history?

Thank you! Yes, a rebase-merge seems apt in this case, but then I'd still like to squash my most recent commit (55cc969) into the first commit (b855b70), since it's a fixup. I'll do that now.

For the regressions, I don't know off-hand, but I could investigate.

@mikewest that seems fine to me if it's done by someone who's not me. :)

**Comment by mikewest** [comment]
> @mikewest that seems fine to me if it's done by someone who's not me. :)

I realize now that this wouldn't help, as it would simply generate a new PR against WPT that would be blocked in the same way this one is. I think we should just land this, and deal with regressions in subsequent CLs, rather than blocking the entire refactoring on specific questions.

Where can I see the list of affected tests? Are we talking about hundreds, or single-digits?

**Comment by zcorpan** [comment]
Taskcluster [failed](https://community-tc.services.mozilla.com/tasks/JQVGrsZNTXeNLlhxBd8h8A/runs/0/logs/https%3A%2F%2Fcommunity-tc.services.mozilla.com%2Fapi%2Fqueue%2Fv1%2Ftask%2FJQVGrsZNTXeNLlhxBd8h8A%2Fruns%2F0%2Fartifacts%2Fpublic%252Flogs%252Flive.log) in the latest run, so there are no results from Firefox or Chrome in wpt.fyi. Per advice from @stephenmcgruer I'll close and reopen to restart and see if that helps...

**Comment by mikewest** [comment]
(I'm running the PR through https://chromium-review.googlesource.com/c/chromium/src/+/2624651 just to see what happens.)

**Comment by zcorpan** [comment]
@stephenmcgruer so the "regressions" (from X/X to 0/0) are actually moved tests (status is `MISSING`). The new tests are in the `generated` folder. So that is expected.

However, in the generated tests, there are a lot of timeouts which I don't see in the previous tests. I'll look into this.

**Comment by mikewest** [comment]
FWIW, the Chromium CQ doesn't like this patch: https://ci.chromium.org/ui/p/chromium/builders/try/linux-rel/584973/test-results?expected=false&exonerated=true&flaky=true. One or two harness failures, many "Error: Failed to query for recorded headers." errors.

**Comment by stephenmcgruer** [comment]
If you're old school like me, the legacy test results viewer for the same run as Mike posted is https://test-results.appspot.com/data/layout_results/linux-rel/584973/blink_web_tests%20%28with%20patch%29/layout-test-results/results.html

**Comment by zcorpan** [comment]
I've identified and fixed an error in `fetch/metadata/resources/record-headers.py`. `json.dumps` throws an exception for `bytes` instances in `response.headers`. I don't know why this worked before, though. Let's see if the tests work as intended with this fix.

**Comment by zcorpan** [comment]
I've noticed another bug. The tests assume that HTTP header values are represented as a string, but now they are wrapped in an array. Maybe this changed in wptserve recently?

```
  ▶ Unexpected subtest result in /fetch/metadata/generated/css-font-face.https.sub.tentative.html:
  │ FAIL [expected PASS] sec-fetch-dest
  │   → assert_equals: expected (string) "font" but got (object) ["font"]
  │ 
  └     at https://web-platform.test:8443/fetch/metadata/generated/css-font-face.https.sub.tentative.html:215:11
```

**Comment by zcorpan** [comment]
An example of calling `print(request.headers)` (without going through `bytes_to_strings`) in `record-headers.py`:

```
{b'host': [b'web-platform.test:8443'], b'connection': [b'keep-alive'], b'sec-ch-ua': [b'"\\"Not\\\\A;Brand";v="99", "Chromium";v="86"'], b'origin': [b'https://web-platform.test:8443'], b'sec-ch-ua-mobile': [b'?0'], b'user-agent': [b'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4232.0 Safari/537.36'], b'accept': [b'*/*'], b'sec-fetch-site': [b'same-origin'], b'sec-fetch-mode': [b'cors'], b'sec-fetch-dest': [b'font'], b'referer': [b'https://web-platform.test:8443/fetch/metadata/generated/css-font-face.https.sub.tentative.html'], b'accept-encoding': [b'gzip, deflate, br'], b'accept-language': [b'en-GB,en-US;q=0.9,en;q=0.8']}
```

**Comment by mikewest** [comment]
https://test-results.appspot.com/data/layout_results/linux-rel/585201/blink_web_tests%20%28with%20patch%29/layout-test-results/results.html looks like `bytes_to_string` fixes most things, leaving us with ~7 timeouts and ~50 failures that look in a quick spot-check to all be the array issue you noted above.

File: fetch/metadata/redirect/redirect-http-upgrade.sub.html [hunk 1]
```diff
@@ -6,62 +6,12 @@
...
-    src: url(http://{{host}}:{{ports[http][0]}}/fetch/api/resources/redirect.py?location=https%3A%2F%2F{{host}}%3A{{ports[https][0]}}%2Ffetch%2Fmetadata%2Fresources%2Frecord-header.py%3Ffile%3Dfont-https-upgrade{{$id:uuid()}});
-  }
-  #fontTest {
-    font-family: myUpgradedFont;
-  }
-</style>
 <body>
-    <div id="fontTest">Upgraded font</div>
     <script>
   let nonce = "{{$id}}";
```
Comment (mikewest) [inline fetch/metadata/redirect/redirect-http-upgrade.sub.html:11]: This is actually broken, as the `{{$id:uuid()}}` call on 12 is gone. But also: we don't need this line anymore. :)

File: fetch/metadata/redirect/multiple-redirect-https-downgrade-upgrade.sub.html [hunk 2]
```diff
@@ -6,68 +6,13 @@
...
-    src: url(https://{{host}}:{{ports[https][0]}}/fetch/api/resources/redirect.py?location=http%3A%2F%2F{{host}}%3A{{ports[http][0]}}%2Ffetch%2Fapi%2Fresources%2Fredirect.py%3Flocation%3Dhttps%253A%252F%252F{{host}}%253A{{ports[https][0]}}%252Ffetch%252Fmetadata%252Fresources%252Frecord-header.py%253Ffile%253Dfont-https-downgrade-upgrade{{$id:uuid()}});
-  }
-  #fontTest {
-    font-family: myDowngradeUpgradeFont;
-  }
-</style>
 <body>
-<div id="fontTest">Downgraded then upgraded font</div>
 <script>
   let nonce = "{{$id}}";
```
Comment (mikewest) [inline fetch/metadata/redirect/multiple-redirect-https-downgrade-upgrade.sub.html:11]: This is actually broken, as the `{{$id:uuid()}}` call on 12 is gone. But also: we don't need this line anymore. :)

**Review by mikewest — COMMENTED** [review]
Also, these two tests have tiny errors:

File: fetch/metadata/redirect/multiple-redirect-https-downgrade-upgrade.sub.html (hunk 2 above)
Comment (zcorpan) [reply fetch/metadata/redirect/multiple-redirect-https-downgrade-upgrade.sub.html]: The test needs `nonce` to be defined to run:

```
promise_test: Unhandled rejection with value: object "ReferenceError: nonce is not defined"
```

However, this was indeed broken. I've fixed it by using `token()` instead.

**Comment by zcorpan** [comment]
I've fixed the array issue (98b856e) and the `nonce` issue (04c4d8b).

File: fetch/metadata/redirect/redirect-http-upgrade.sub.html [hunk 3]
```diff
@@ -7,8 +7,8 @@
 <script src=/common/security-features/resources/common.sub.js></script>
 <script src=/common/utils.js></script>
 <body>
-    <script>
-  let nonce = "{{$id}}";
+<script>
+  let nonce = token();
```
Comment (mikewest) [inline fetch/metadata/redirect/redirect-http-upgrade.sub.html:11]: I don't think `nonce` is used anywhere in this file now (nor in `multiple-redirect-https-downgrade-upgrade.sub.html`). Is it used as a global from within `RunCommonRedirectTests()`?

Comment (zcorpan) [reply fetch/metadata/redirect/redirect-http-upgrade.sub.html:11]: Yes:

https://github.com/web-platform-tests/wpt/blob/master/fetch/metadata/resources/redirectTestHelper.sub.js#L74

That's not great, but this PR doesn't seem like the best place to fix it.

Comment (mikewest) [reply fetch/metadata/redirect/redirect-http-upgrade.sub.html:11]: Got it. In that case, I agree!

**Comment by mikewest** [comment]
Thanks! I'll throw this update at Chromium's trybots to see what they say: https://chromium-review.googlesource.com/c/chromium/src/+/2624651.

**Comment by zcorpan** [comment]
Azure Pipelines check failed:

```
tools.wpt.run.WptrunError: Unable to locate safaridriver binary
```
https://dev.azure.com/web-platform-tests/wpt/_build/results?buildId=59715&view=logs&j=bf8f5026-3d1f-5955-26be-17f9e1ef0bb3&t=f2f2a672-3250-5564-df22-32ed9c3bfd05&l=28

@stephenmcgruer any advice?

**Comment by stephenmcgruer** [comment]
> @stephenmcgruer any advice?

:(. Let me take a look... also ping @gsnedders as an FYI.

**Comment by stephenmcgruer** [comment]
Oh, fun: https://dev.azure.com/web-platform-tests/wpt/_build/results?buildId=59715&view=logs&j=bf8f5026-3d1f-5955-26be-17f9e1ef0bb3&t=fb374013-b51a-5838-c2f2-591e11e493ab

Installing STP errored, but didn't turn the step red -_-.

```
Error: Calling brew cask install is disabled! Use brew install [--cask] instead.
Finishing: Install Safari Technology Preview
```

I'll see if I can get this fixed asap.

**Comment by stephenmcgruer** [comment]
> I'll see if I can get this fixed asap.

I've sent https://github.com/web-platform-tests/wpt/pull/27226 which I believe should fix Azure Pipelines.

**Comment by mikewest** [comment]
> Thanks! I'll throw this update at Chromium's trybots to see what they say: https://chromium-review.googlesource.com/c/chromium/src/+/2624651.

Chromium's CQ is happier, but still points out a few issues: https://test-results.appspot.com/data/layout_results/linux_layout_tests_composite_after_paint/35270/blink_web_tests%20%28with%20patch%29/layout-test-results/results.html.

**Comment by zcorpan** [comment]
Thanks, @mikewest. Work-in-progress analysis:

## `fetch/metadata/generated/audioworklet.https.sub.html`

This is a new test with a single subtest failing. Looks like it can be a bug in Chromium. TODO file bug.


## `fetch/metadata/generated/cors-preflight.https.sub.html`

This test appears to be broken. It looks like it's intended to replace `fetch/metadata/fetch-preflight.https.sub.any.html` but it's not covering the serviceworker, sharedworker, or worker tests, and it's failing all subtests. `record-headers.py` doesn't return the same headers in the preflight as `echo-as-json.py` does, which may be the issue. In particular:
```
  # This condition avoids false positives from CORS preflight checks, where the
  # request under test may be followed immediately by a request to the same URL
  # using a different HTTP method.
  if b'requireOPTIONS' in request.GET and request.method != b'OPTIONS':
      return
```
vs `echo-as-json.py`:
```
    # If we're in a preflight, verify that `Sec-Fetch-Mode` is `cors`.
    if request.method == u'OPTIONS':
        if request.headers.get(b"sec-fetch-mode") != b"cors":
            return (403, b"Failed"), [], body

        headers.append((b"Access-Control-Allow-Methods", b"*"))
        headers.append((b"Access-Control-Allow-Headers", b"*"))
```
I suggest we move the `cors-preflight` test into a separate PR so it can be fixed without blocking the rest of this PR.

## `fetch/metadata/generated/css-images.https.sub.tentative.html`

This is a new test.

```
  async_test((t) => {
    const key = 'f13118cb-656b-49ab-b45d-7e8c5d45b292';
    const url = makeRequestURL(key, ['httpsOrigin', 'httpOrigin', 'httpsOrigin']);

    declarations.push(`background-image: url("${url}");`);

    whenIframeReady
      .then(() => retrieve(key))
      .then((headers) => {
          assert_own_property(headers, 'sec-fetch-site');
          assert_equals(headers['sec-fetch-site'], 'cross-site');
        })
        .then(t.step_func_done(), (error) => t.unreached_func());
  }, 'background-image sec-fetch-site - HTTPS downgrade-upgrade');
```
This somehow makes Chrome request the following URL, which fails because port 8000 doesn't do a TLS handshake:
```
https://web-platform.test:8000/fetch/api/resources/redirect.py?location=https%3A%2F%2Fweb-platform.test%3A8443%2Ffetch%2Fmetadata%2Fresources%2Frecord-headers.py%3Fkey%3Df13118cb-656b-49ab-b45d-7e8c5d45b292&location=http%3A%2F%2Fweb-platform.test%3A8000%2Ffetch%2Fapi%2Fresources%2Fredirect.py%3Flocation%3Dhttps%253A%252F%252Fweb-platform.test%253A8443%252Ffetch%252Fmetadata%252Fresources%252Frecord-headers.py%253Fkey%253Df13118cb-656b-49ab-b45d-7e8c5d45b292&count=1
```
But `makeRequestURL` (in `helper.sub.js`) can't produce the above URL; is uses port 8000 for `http` only. Does Chrome change the scheme in this request from `http` to `https`, leaving the port alone?

## Remaining TODO

**Comment by zcorpan** [comment]
OK, I've now gone through all tests listed in "Regressions" in https://test-results.appspot.com/data/layout_results/linux_layout_tests_composite_after_paint/35270/blink_web_tests%20%28with%20patch%29/layout-test-results/results.html (see the previous comment).

**Comment by gsnedders** [comment]
> > @stephenmcgruer any advice?
> 
> :(. Let me take a look... also ping @gsnedders as an FYI.

this was fixed in https://github.com/web-platform-tests/wpt/pull/27477 FYI

**Comment by zcorpan** [comment]
Rebased again. https://github.com/web-platform-tests/wpt/pull/25247#issuecomment-762755868 lists a few TODOs (which could be follow-ups), though I'm not currently working on this.

**Comment by foolip** [comment]
@jugglinmike I've rebased this branch to resolve conflicts, can you enable "Allow edits and access to secrets by maintainers" so I can push those? (@zcorpan asked me)

**Comment by foolip** [comment]
https://wpt.fyi/results/fetch/metadata?label=pr_head&max-count=1&pr=25247 show the results.

Everything looks OK except https://wpt.fyi/results/fetch/metadata/generated/appcache-manifest.https.sub.html?label=pr_head&max-count=1&pr=25247 where there's a harness error:

> Test named "sec-fetch-site - Same origin" passed a function to `async_test` that returned a value. Consider using `promise_test` instead when using Promises or async/await.

This should be fixed by just not returning a promise, it's already chained correctly to pass or fail the async test.

Stability jobs are failing on /webaudio/the-audio-api/the-mediaelementaudiosourcenode-interface/no-cors.https.html because fetch/api/resources/redirect.py was updated, but the test does a regular GET request and I presume was already flaky. I'll admin merge to ignore this failure.

**Comment by foolip** [comment]
https://wpt.fyi/results/fetch/metadata/generated/appcache-manifest.https.sub.html?label=pr_head&max-count=1&pr=25247 looks better now, no longer a harness error.

**Comment by zcorpan** [comment]
Thank you @foolip !

**Pull request merged**